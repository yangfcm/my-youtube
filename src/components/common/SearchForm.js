import React from "react";
import { withRouter } from "react-router-dom";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

class SearchForm extends React.Component {
  state = {
    search: "",
  };

  handleSearch = (e) => {
    e.preventDefault();
    if (this.state.search.trim() === "") {
      return;
    }
    this.props.history.push(`/results?q=${this.state.search}`);
    this.setState({
      search: "",
    });
  };

  handleInput = (e) => {
    this.setState({
      search: e.target.value,
    });
  };

  render() {
    return (
      <form className="d-flex" onSubmit={this.handleSearch}>
        <input
          type="text"
          placeholder="Search video..."
          className="form-control"
          name="search"
          value={this.state.search}
          onChange={this.handleInput}
        />
        <button className="btn btn-outline-dark mx-2" type="submit">
          <FontAwesomeIcon icon="search" />
        </button>
      </form>
    );
  }
}

export default withRouter(SearchForm);
