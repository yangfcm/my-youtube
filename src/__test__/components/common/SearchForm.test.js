import React from "react";
import { shallow } from "enzyme";
import SearchForm from "components/common/SearchForm";

describe("Test SearchForm component", () => {
  let wrapper;
  const mockPush = jest.fn();
  beforeEach(() => {
    wrapper = shallow(
      <SearchForm.WrappedComponent history={{ push: mockPush }} />
    );
  });
  afterEach(() => {
    jest.clearAllMocks();
    wrapper.unmount();
  });

  it("should render component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render form", () => {
    expect(wrapper.find("form").exists()).toBe(true);
    expect(wrapper.find("input").exists()).toBe(true);
    expect(wrapper.find("button").exists()).toBe(true);
  });

  it("should input value in the form", () => {
    const input = wrapper.find("input");
    const inputValue = "google dev";
    input.prop("onChange")({ target: { value: inputValue } });
    expect(wrapper.state("search")).toBe(inputValue);
  });

  it("should redirect to results page with search keyword when clicking search button with some input", () => {
    const inputValue = "google dev";
    const input = wrapper.find("input");
    const form = wrapper.find("form");
    input.prop("onChange")({ target: { value: inputValue } });
    form.prop("onSubmit")({ preventDefault: () => {} });
    expect(mockPush).toHaveBeenCalled();
    expect(mockPush).toHaveBeenCalledWith(`/results?q=${inputValue}`);
    expect(wrapper.state("search")).toBe("");
  });

  it("should not redirect to results page when clicking search button with no input", () => {
    const form = wrapper.find("form");
    wrapper.setState({ search: "" });
    form.prop("onSubmit")({ preventDefault: () => {} });
    expect(mockPush).toHaveBeenCalledTimes(0);
  });
});
