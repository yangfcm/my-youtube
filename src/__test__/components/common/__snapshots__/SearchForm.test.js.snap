// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test SearchForm component should render component 1`] = `
<form
  className="d-flex"
  onSubmit={[Function]}
>
  <input
    className="form-control"
    name="search"
    onChange={[Function]}
    placeholder="Search video..."
    type="text"
    value=""
  />
  <button
    className="btn btn-outline-dark mx-2"
    type="submit"
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon="search"
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
  </button>
</form>
`;
